src/pre_process/m_model.fpp

```diff
@@ -549,4 +554,697 @@ contains
+    !! @param spacing                      Dimensions of the current levelset cell
+    !! @param interpolated_boundary_v      Output all the boundary vertices of the interpolated 3D model
+    !! @param total_vertices               Total number of vertices after interpolation
+    subroutine f_interpolate_3D(model, spacing, interpolated_boundary_v, total_vertices)
+        t_vec3, intent(in) :: spacing
+        type(t_model), intent(in) :: model
+        real(kind(0d0)), allocatable, intent(inout), dimension(:, :) :: interpolated_boundary_v
+        integer, intent(out) :: total_vertices
+
+        integer :: i, j, k, num_triangles, num_segments
+        real(kind(0d0)) :: x1, y1, z1, x2, y2, z2, x3, y3, z3
+        real(kind(0d0)) :: edge_length, del_x, del_y, del_z, cell_width
+        real(kind(0d0)) :: area_xy, area_xz, area_yz, cell_area, tri_area
+        real(kind(0d0)) :: u, v, w !< Barycentric coordinates
+        integer :: num_inner_vertices
+        real(kind(0d0)), allocatable :: temp_boundary_v(:, :)
+
+        ! Number of triangles in the model
+        num_triangles = model%ntrs
+        cell_width = minval(spacing)
+
+        ! Find the minimum surface area
+        area_xy = spacing(1)*spacing(2)
+        area_xz = spacing(1)*spacing(3)
+        area_yz = spacing(2)*spacing(3)
+        cell_area = minval((/area_xy, area_xz, area_yz/))
+        num_inner_vertices = 0
+
+        ! Calculate the total number of vertices including interpolated ones
+        total_vertices = 0
+        do i = 1, num_triangles
+            do j = 1, 3
+                ! Get the coordinates of the two vertices of the current edge
+                x1 = model%trs(i)%v(j, 1)
+                y1 = model%trs(i)%v(j, 2)
+                z1 = model%trs(i)%v(j, 3)
+                ! Next vertex in the triangle (cyclic)
+                x2 = model%trs(i)%v(mod(j, 3) + 1, 1)
+                y2 = model%trs(i)%v(mod(j, 3) + 1, 2)
+                z2 = model%trs(i)%v(mod(j, 3) + 1, 3)
+
+                ! Compute the length of the edge
+                edge_length = dsqrt((x2 - x1)**2 + (y2 - y1)**2 + (z2 - z1)**2)
+
+                ! Determine the number of segments
+                if (edge_length > cell_width) then
+                    num_segments = Ifactor_3D*ceiling(edge_length/cell_width)
+                else
+                    num_segments = 1
+                end if
+
+                ! Each edge contributes num_segments vertices
+                total_vertices = total_vertices + num_segments + 1
+            end do
+
+            ! Add vertices inside the triangle
+            x1 = model%trs(i)%v(1, 1)
+            y1 = model%trs(i)%v(1, 2)
+            z1 = model%trs(i)%v(1, 3)
+            x2 = model%trs(i)%v(2, 1)
+            y2 = model%trs(i)%v(2, 2)
+            z2 = model%trs(i)%v(2, 3)
+            x3 = model%trs(i)%v(3, 1)
+            y3 = model%trs(i)%v(3, 2)
+            z3 = model%trs(i)%v(3, 3)
```

Comment:
There's an existing derived type, `t_triangle`, that would these assignments conform to.